### PR TITLE
crypto_api: ecc allocation api

### PR DESCRIPTION
**Pull request by ldts**
For **Elliptic Curve**, the cryptographic api can fallback to its software
operation instead of failing due to the lack of hardware support.
    
The relevant code can be see seen in key allocation functions for 
the [public](https://github.com/OP-TEE/optee_os/blob/c4cab13e43f8170575612892f66bdbc8c9e67285/core/crypto/crypto.c#L709) and [private](https://github.com/OP-TEE/optee_os/blob/c4cab13e43f8170575612892f66bdbc8c9e67285/core/crypto/crypto.c#L727) keys

crypto_api/acipher/ecc.c (i.e [pair]( https://github.com/OP-TEE/optee_os/blob/c4cab13e43f8170575612892f66bdbc8c9e67285/core/drivers/crypto/crypto_api/acipher/ecc.c#L348) ) however does not pass the key type to 
the relevant driver and therefore the driver can not take the correct
action at allocation time.
    
This commit addresses that limitation and fixes 3.21.0 regression_4006 
for the **Versal ACAP** platform and any **NXP SE05X enabled** platform.
` regression_4006.43 Asym Crypto case 426 algo 0x80000046 line 373`

### CODE REVIEW

File: core/drivers/crypto/crypto_api/include/drvcrypt_acipher.h [hunk 1]
```diff
@@ -149,10 +149,11 @@ struct drvcrypt_ecc_ed {
  */
 struct drvcrypt_ecc {
 	/* Allocates the ECC keypair */
-	TEE_Result (*alloc_keypair)(struct ecc_keypair *key, size_t size_bits);
+	TEE_Result (*alloc_keypair)(struct ecc_keypair *key, size_t size_bits,
+				    uint32_t type);
```
Comment (jenswi-linaro) [inline core/drivers/crypto/crypto_api/include/drvcrypt_acipher.h:153]: We may want to swap the order of `size_bits` and `type` to match `drvcrypt_asym_alloc_ecc_keypair()`

**Comment by clementfaure** [comment]
With @jenswi-linaro comment, `Acked-by: Clement Faure <clement.faure@nxp.com>` for 72c93de7a5976505a8019e7cefe07bdd46d65fd3

**Comment by ldts** [comment]
> With @jenswi-linaro comment, `Acked-by: Clement Faure <clement.faure@nxp.com>` for [72c93de](https://github.com/OP-TEE/optee_os/commit/72c93de7a5976505a8019e7cefe07bdd46d65fd3)

hi @clementfaure ,at the time of your comment the swap that @jenswi-linaro suggested is already done. Just wanted to double check that you are ok with it.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commit 
"crypto_api: acipher: ecc key allocation api, pass the key type" with nitpicking changes in commit message: s/api/API/ + s/IP driver/backend driver/ (or device driver).

`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
"crypto: se050: ecc: allow software fallback on key allocation requests" and
"crypto: versal: ecc: allow software fallback on key allocation" with a suggestion.

File: core/drivers/crypto/se050/core/ecc.c [hunk 2]
```diff
@@ -783,6 +790,7 @@ static struct drvcrypt_ecc driver_ecc = {
 
 static TEE_Result ecc_init(void)
 {
+	/* This driver only supports ECDH/ECDSA */
```
[the remark this answers is not in the document]
Comment (ldts) [reply core/drivers/crypto/se050/core/ecc.c:793]: sure I can do that. makes sense to me.

**Comment by clementfaure** [comment]
Yes, you still have my ack on this patch

**Review by etienne-lms — COMMENTED** [review]
In commit message for "crypto_api: acipher: ecc key allocation API, pass the key type":
```patch
-For Elliptic Curve, the cryptographic api can fallback to its software
+For Elliptic Curve, the cryptographic API can fallback to its software
```
Also can replace my A-b tag with my R-b tag (nitpicking again :-)

In commit "crypto: versal: ecc: allow software fallback on key allocation":
s/pair_ops/keypair_ops/  (see CI build error report)

**Comment by ldts** [comment]
tags added @jforissier . thanks all for the comments and review.

**Pull request merged**